### ISSUE TITLE
doubts about the operation of the zigbee stack (TZ-139)

### ISSUE THREAD

**Issue by diazmanuel** [espressif/esp-zigbee-sdk#32]
hello!!

I understand that the zigbee stack is usually a problem since once it fills up, problems begin to occur

Looking at the examples, notice that each time the device is restarted, the bindings are created again. One of the queries is if, when creating these bindings, an internal check is made to see if said binding already existed.

If this check is not done, is there any method to delete the different components of the stack such as bindings, report configurations and the network to which it was connected, or to know the status of the zigbee stack?

I have noticed some strange behaviors that were solved by erasing the flash memory of the chip, maybe it is due to the zigbee stack

I take the opportunity to make some other queries, from time to time I receive signal 50 (ESP_ZB_NLME_STATUS_INDICATION ) in the connection handler but I don't know what it means exactly, after this signal I receive a signal to reconnect to the network and everything starts to work correctly.

Finally, I would like to know if it is possible to declare a cluster as client and server simultaneously, I tried implementing it with an or between the role definitions but it does not work correctly, then I declared it twice on the same endpoint (one as server and the other as client) but the reports do not work correctly, specifically it has problems with the periodic reports, if more information is needed for this case I can upload the code, I am working with a system that has at least 8 endpoints which could also be the cause of this malfunction

Thank you in advance and sorry for mixing several topics in the same issue

**Comment by likunqiao097304** [comment]
> If this check is not done, is there any method to delete the different components of the stack such as bindings, report configurations and the network to which it was connected, or to know the status of the zigbee stack

Are you looking for something like `esp_zb_factory_reset` to remove history data or `esp_zb_nvram_erase_at_start` to erase nvram data which contains those information every time you restart?



> I have noticed some strange behaviors that were solved by erasing the flash memory of the chip, maybe it is due to the zigbee stack

Could you specify what exactly strange behaviors you saw?




> I would like to know if it is possible to declare a cluster as client and server simultaneously

You'd like to have a one API to declare both server and client at one endpoint? For the issue about the report, could you give us more detail information about how is your network topology and binding, reporting set up? Also it would be nice to share us some code, We will try to solve it together

**Comment by diazmanuel** [comment]
@likunqiao097304 
The first part comes as a question of the bindings table, in the zboss library at a certain point the table in charge of the bindings was filled, for example every time the device was restarted, the binding table was filled with the same bind. For this, the zboss library allowed you to read said table to check if the binding you were wanting to enter already existed, and in that case avoid adding it again.

Regarding the strange behavior, I could not replicate it because once the flash was deleted and the device was rewritten, it began to work correctly, but what I could notice was that the reports stopped working, and also it did not allow me to add more than one cluster to the same endpoint. the chip was directly restarted by the simple declaration of the cluster, below I leave the code in case you want to analyze it

what i mean by declaring a cluster as both server and client is as bellow
this way 
    esp_zb_cluster_list_add_custom_cluster(esp_zb_cluster_list, FFFF, ESP_ZB_ZCL_CLUSTER_CLIENT_ROLE );
    esp_zb_cluster_list_add_custom_cluster(esp_zb_cluster_list, FFFF, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);
or this
    esp_zb_cluster_list_add_custom_cluster(esp_zb_cluster_list, FFFF, ESP_ZB_ZCL_CLUSTER_CLIENT_ROLE | ESP_ZB_ZCL_CLUSTER_SERVER_ROLE );

The purpose of this is that this device we call B receives a report from A to said cluster and it must replicate the report to C. This is so because C must be sure that the report from A reached B correctly (this is important since otherwise A could report to both B and C)

The topology is as follows: there is a gateway, a control device with multiple outputs (let's call it Switch), up to 6 thermostats and 1 boiler. The code corresponds to the switch.
So the thermostats report through the cluster on off to the different endpoints of the switch and this activates the outputs according to which thermostat makes the report. If there is an output activated, the Switch reports to the boiler through the on-off cluster so that it is activated. In turn, if the boiler detects an error, it reports said error to the switch through a custom cluster (FFFF) and the switch in turn reports to each of the thermostats through said cluster. In parallel each status of the outputs and the status of the boiler are reported from the switch to the gateway. here is the code of the switch:
```
#include "taskZigbee.h"



static const char *TAG = "taskZigbee";

struct cluster_data_t cluster_data;


typedef struct zdo_info_ctx_s {
    uint8_t src_endp;
    uint16_t cluster_id;
    //esp_zb_ieee_addr_t long_addr;
} zdo_info_user_ctx_t;



static void bdb_start_top_level_commissioning_cb(uint8_t mode_mask)
{
    ESP_ERROR_CHECK(esp_zb_bdb_start_top_level_commissioning(mode_mask));
}

static void bind_cb(esp_zb_zdp_status_t zdo_status, void *user_ctx)
{
    if (zdo_status == ESP_ZB_ZDP_STATUS_SUCCESS) {
        ESP_LOGI(TAG, "bind_cb status:%d, clsuter %d, src_endp %d", zdo_status,((zdo_info_user_ctx_t *)user_ctx)->cluster_id,((zdo_info_user_ctx_t *)user_ctx)->src_endp);
        /* configure report attribute command */
        static esp_zb_zcl_config_report_cmd_t report_cmd;
        static esp_zb_zcl_config_report_cmd_t report_cmd_2;
        if(((zdo_info_user_ctx_t *)user_ctx)->cluster_id == 0xFFFF){
            int16_t report_change=1;
            report_cmd.zcl_basic_cmd.dst_addr_u.addr_short=esp_zb_get_short_address();
            report_cmd.zcl_basic_cmd.dst_endpoint = 1;
            report_cmd.zcl_basic_cmd.src_endpoint = ((zdo_info_user_ctx_t *)user_ctx)->src_endp;
            report_cmd.address_mode = ESP_ZB_APS_ADDR_MODE_16_ENDP_PRESENT;
            report_cmd.clusterID = ((zdo_info_user_ctx_t *)user_ctx)->cluster_id;
            report_cmd.attributeID = 0x05;
            report_cmd.attrType = ESP_ZB_ZCL_ATTR_TYPE_S16;
            report_cmd.min_interval = 0;
            report_cmd.max_interval = 300;
            report_cmd.reportable_change = (void*)&report_change;
            esp_zb_zcl_config_report_cmd_req(&report_cmd);
        }
        if(((zdo_info_user_ctx_t *)user_ctx)->cluster_id == 0x0006){
            bool report_change_2=1;
            report_cmd_2.zcl_basic_cmd.dst_addr_u.addr_short=esp_zb_get_short_address();
            report_cmd_2.zcl_basic_cmd.dst_endpoint = 1;
            report_cmd_2.zcl_basic_cmd.src_endpoint = ((zdo_info_user_ctx_t *)user_ctx)->src_endp;
            ESP_LOGI(TAG,"%d    %d",((zdo_info_user_ctx_t *)user_ctx)->cluster_id,((zdo_info_user_ctx_t *)user_ctx)->src_endp);
            report_cmd_2.address_mode = ESP_ZB_APS_ADDR_MODE_16_ENDP_PRESENT;
            report_cmd_2.clusterID = ((zdo_info_user_ctx_t *)user_ctx)->cluster_id;
            report_cmd_2.attributeID = ESP_ZB_ZCL_ATTR_ON_OFF_ON_OFF_ID;
            report_cmd_2.attrType = ESP_ZB_ZCL_ATTR_TYPE_BOOL;
            report_cmd_2.min_interval = 0;
            report_cmd_2.max_interval = 0xFFFF;
            report_cmd_2.reportable_change = (void*)&report_change_2;
            esp_zb_zcl_config_report_cmd_req(&report_cmd_2);
        }

    }
}


static void esp_zb_dev_reporting_cb(esp_zb_zcl_addr_t *addr, uint8_t src_endpoint, uint8_t dst_endpoint, uint16_t cluster_id,
                                    uint16_t attr_id, esp_zb_zcl_attr_type_t attr_type, void *value)
{

        if(cluster_id==0x0006){
            esp_zb_zcl_set_attribute_val(dst_endpoint,cluster_id,ESP_ZB_ZCL_CLUSTER_SERVER_ROLE,attr_id,value);   

        }
        else if(cluster_id==0xFFFF){
            esp_zb_zcl_set_attribute_val(dst_endpoint,cluster_id,ESP_ZB_ZCL_CLUSTER_SERVER_ROLE,attr_id,value);
        }  
    

}
static void ieee_cb(esp_zb_zdp_status_t zdo_status, esp_zb_ieee_addr_t ieee_addr, void *user_ctx)
{
    
    if (zdo_status == ESP_ZB_ZDP_STATUS_SUCCESS) {
        static zdo_info_user_ctx_t info_ctx[8];

        // memcpy(&(on_off_light.ieee_addr), ieee_addr, sizeof(esp_zb_ieee_addr_t));
        ESP_LOGI(TAG, "ieee_cb address is:%02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x",
                 ieee_addr[7], ieee_addr[6], ieee_addr[5], ieee_addr[4],
                 ieee_addr[3], ieee_addr[2], ieee_addr[1], ieee_addr[0]);
        // /* bind the on-off light to on-off switch */
        esp_zb_zdo_bind_req_param_t bind_req;
        esp_zb_get_long_address(bind_req.src_address);

        bind_req.dst_addr_mode = ESP_ZB_ZDO_BIND_DST_ADDR_MODE_64_BIT_EXTENDED;
        memcpy(bind_req.dst_address_u.addr_long, ieee_addr, sizeof(esp_zb_ieee_addr_t));

        ESP_LOGI(TAG, "local address is:%02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x",
                bind_req.src_address[7], bind_req.src_address[6], bind_req.src_address[5],
               bind_req.src_address[4], bind_req.src_address[3], bind_req.src_address[2],
                bind_req.src_address[1], bind_req.src_address[0]);

        bind_req.dst_endp = 1;
        bind_req.req_dst_addr = esp_zb_get_short_address();


        info_ctx[0].src_endp = ZIGBEE_BOILER_ENDPOINT;
        info_ctx[0].cluster_id = 0xFFFF;
        bind_req.src_endp = info_ctx[0].src_endp;
        bind_req.cluster_id = info_ctx[0].cluster_id;
        esp_zb_zdo_device_bind_req(&bind_req, bind_cb, (void *) &(info_ctx[0]));

        info_ctx[1].src_endp = ZIGBEE_BOILER_ENDPOINT;
        info_ctx[1].cluster_id = 0x0006;
        bind_req.src_endp = info_ctx[1].src_endp;
        bind_req.cluster_id = info_ctx[1].cluster_id;
        esp_zb_zdo_device_bind_req(&bind_req, bind_cb, (void *) &(info_ctx[1]));

        for(int i=0;i<ZONAS;i++){
            info_ctx[2+i].cluster_id = 0x0006;    
            bind_req.cluster_id = info_ctx[2+i].cluster_id;
            info_ctx[2+i].src_endp = ZIGBEE_ZONAS_ENDPOINT_OFFSET+i;
            bind_req.src_endp = info_ctx[2+i].src_endp;
            esp_zb_zdo_device_bind_req(&bind_req, bind_cb, (void *) &(info_ctx[i+2]));
        }

    }
}


void esp_zb_app_signal_handler(esp_zb_app_signal_t *signal_struct)
{
    
    uint32_t *p_sg_p       = signal_struct->p_app_signal;
    esp_err_t err_status = signal_struct->esp_err_status;
    esp_zb_app_signal_type_t sig_type = *p_sg_p;
    switch (sig_type) {
    case ESP_ZB_ZDO_SIGNAL_SKIP_STARTUP:
        ESP_LOGI(TAG, "Zigbee stack initialized");
        esp_zb_bdb_start_top_level_commissioning(ESP_ZB_BDB_MODE_INITIALIZATION);
        AppPerifericosHandlerOutput(EVENT_TASK_PERIFERICOS_HANDLER_OUTPUT_LED_ZIGBEE,true,0,true);
        break;
    case ESP_ZB_BDB_SIGNAL_DEVICE_FIRST_START:
    case ESP_ZB_BDB_SIGNAL_DEVICE_REBOOT:
        if (err_status == ESP_OK) {
            ESP_LOGI(TAG, "Start network steering");
            esp_zb_bdb_start_top_level_commissioning(ESP_ZB_BDB_MODE_NETWORK_STEERING);
        } else {
            /* commissioning failed */
            ESP_LOGW(TAG, "Failed to initialize Zigbee stack (status: %d)", err_status);
            esp_zb_bdb_start_top_level_commissioning(ESP_ZB_BDB_MODE_NETWORK_STEERING);
        }
        
        break; 
    case ESP_ZB_BDB_SIGNAL_STEERING:
        if (err_status == ESP_OK) {
            AppPerifericosHandlerOutput(EVENT_TASK_PERIFERICOS_HANDLER_OUTPUT_LED_ZIGBEE,true,0,false);
            esp_zb_ieee_addr_t extended_pan_id;
            esp_zb_get_extended_pan_id(extended_pan_id);
            ESP_LOGI(TAG, "Joined network successfully (Extended PAN ID: %02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x, PAN ID: 0x%04hx, Channel:%d)",
                     extended_pan_id[7], extended_pan_id[6], extended_pan_id[5], extended_pan_id[4],
                     extended_pan_id[3], extended_pan_id[2], extended_pan_id[1], extended_pan_id[0],
                     esp_zb_get_pan_id(), esp_zb_get_current_channel());
                     
            esp_zb_zdo_ieee_addr_req_param_t ieee_req;
            ieee_req.addr_of_interest = 0;
            ieee_req.dst_nwk_addr = 0;
            ieee_req.request_type = 0;
            ieee_req.start_index = 0;
            esp_zb_zdo_ieee_addr_req(&ieee_req, ieee_cb, NULL);
        } else {
            ESP_LOGI(TAG, "Network steering was not successful (status: %d)", err_status);
            esp_zb_scheduler_alarm((esp_zb_callback_t)bdb_start_top_level_commissioning_cb, ESP_ZB_BDB_MODE_NETWORK_STEERING, 1000);
        }
        break;
    case ESP_ZB_ZDO_SIGNAL_LEAVE:
        esp_zb_zdo_signal_leave_params_t *leave_params = (esp_zb_zdo_signal_leave_params_t *)esp_zb_app_signal_get_params(p_sg_p);
        if (leave_params->leave_type == ESP_ZB_NWK_LEAVE_TYPE_RESET) {
            ESP_LOGI(TAG, "Reset device");
        }
        break;
    default:
        ESP_LOGI(TAG, "ZDO signal: %d, status: %d", sig_type, err_status);
        break;
    }
}
void attr_cb(uint8_t status, uint8_t endpoint, uint16_t cluster_id, uint16_t attr_id, void *value)
{
    ESP_LOGI(TAG,"status %d,endpoint %d,cluster_id %d,attr_id %d,value %lu,",status,endpoint,cluster_id,attr_id,*(uint32_t *)value);
}



void ZigbeeHandlertTask(void * parameter){

    /* initialize Zigbee stack with Zigbee end-device config */
    esp_zb_cfg_t zb_nwk_cfg = ESP_ZB_ZED_CONFIG();
    esp_zb_init(&zb_nwk_cfg);

    esp_zb_on_off_cluster_cfg_t on_off_cfg;

    on_off_cfg.on_off = ESP_ZB_ZCL_ON_OFF_ON_OFF_DEFAULT_VALUE;
    esp_zb_attribute_list_t *esp_zb_on_off_cluster = esp_zb_on_off_cluster_create(&on_off_cfg);

    esp_zb_ep_list_t *esp_zb_ep_list = esp_zb_ep_list_create();

    esp_zb_cluster_list_t *esp_zb_cluster_list = esp_zb_zcl_cluster_list_create();



    
    esp_zb_attribute_list_t *FFFF = esp_zb_zcl_attr_list_create(0xFFFF);

    esp_zb_custom_cluster_add_custom_attr(FFFF, 0x05, ESP_ZB_ZCL_ATTR_TYPE_S16, ESP_ZB_ZCL_ATTR_ACCESS_READ_ONLY | ESP_ZB_ZCL_ATTR_ACCESS_REPORTING,&cluster_data.FFFE_away);



// Before erasing the flash memory, it was in this place where the failure and restart of the chip occurred. If I added more than one cluster to that endpoint, the device would crash.
 
    esp_zb_cluster_list_add_custom_cluster(esp_zb_cluster_list, FFFF, ESP_ZB_ZCL_CLUSTER_CLIENT_ROLE );
    esp_zb_cluster_list_add_custom_cluster(esp_zb_cluster_list, FFFF, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);
    esp_zb_cluster_list_add_on_off_cluster(esp_zb_cluster_list, esp_zb_on_off_cluster, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);
    esp_zb_ep_list_add_ep(esp_zb_ep_list, esp_zb_cluster_list, ZIGBEE_BOILER_ENDPOINT, 260, ESP_ZB_HA_CUSTOM_ATTR_DEVICE_ID);


    esp_zb_cluster_list = esp_zb_zcl_cluster_list_create();
    esp_zb_cluster_list_add_on_off_cluster(esp_zb_cluster_list, esp_zb_on_off_cluster, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);
    esp_zb_ep_list_add_ep(esp_zb_ep_list, esp_zb_cluster_list, ZIGBEE_WATER_PUMPS_ENDPOINT, 260, ESP_ZB_HA_CUSTOM_ATTR_DEVICE_ID);

    for(int i=0;i<ZONAS;i++){
        esp_zb_cluster_list = esp_zb_zcl_cluster_list_create();
        esp_zb_cluster_list_add_on_off_cluster(esp_zb_cluster_list, esp_zb_on_off_cluster, ESP_ZB_ZCL_CLUSTER_CLIENT_ROLE );
        esp_zb_cluster_list_add_on_off_cluster(esp_zb_cluster_list, esp_zb_on_off_cluster, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);
        esp_zb_ep_list_add_ep(esp_zb_ep_list, esp_zb_cluster_list, ZIGBEE_ZONAS_ENDPOINT_OFFSET+i, 260, ESP_ZB_HA_CUSTOM_ATTR_DEVICE_ID);
    }


    esp_zb_device_register(esp_zb_ep_list);
    esp_zb_device_add_set_attr_value_cb(attr_cb);
    esp_zb_device_add_report_attr_cb(esp_zb_dev_reporting_cb);

    esp_zb_set_primary_network_channel_set(ESP_ZB_PRIMARY_CHANNEL_MASK);


    ESP_ERROR_CHECK(esp_zb_start(false));
    esp_zb_main_loop_iteration();
    
}
uint8_t  InitTaskZigbee(){

esp_zb_platform_config_t config = {
        .radio_config = ESP_ZB_DEFAULT_RADIO_CONFIG(),
        .host_config = ESP_ZB_DEFAULT_HOST_CONFIG(),
    };
    ESP_ERROR_CHECK(esp_zb_platform_config(&config));
    if(xTaskCreate(ZigbeeHandlertTask, "zigbeeHandlertTask", ZIGBEE_HANDLER_TASK_SIZE, NULL, ZIGBEE_HANDLER_TASK_PRIORITY, NULL)!=pdPASS) return false;

    ESP_LOGI(TAG, "Zigbee Handler Task Created");

    return true;

}


```

**Comment by xieqinan** [comment]
Hi,
Let's start by discussing the bind query issue first. If you need the binding table request feature, we can provide it. However, please note that the cost between the bind request and the binding table request may be similar.

> but what I could notice was that the reports stopped working, and also it did not allow me to add more than one cluster to the same endpoint.

Could you please provide the log of the chip related to the issue? I have executed your code on my end, and I haven't encountered any problems. Have you made any new discoveries or observations regarding this matter?

> esp_zb_cluster_list_add_custom_cluster(esp_zb_cluster_list, FFFF, ESP_ZB_ZCL_CLUSTER_CLIENT_ROLE );
> esp_zb_cluster_list_add_custom_cluster(esp_zb_cluster_list, FFFF, ESP_ZB_ZCL_CLUSTER_SERVER_ROLE);

The way is right.

**Comment by diazmanuel** [comment]
hi, @xieqinan,
I am interested in bring the bindings table to confirm if there are duplicates, that is, if a binding was created 2 times, since the binding is executed every time the device is restarted. In case duplicates are generated, I would use this table to decide whether or not to create the binding (if it already exists, I would not create it again)

I think I found the problem, although it is a bit difficult to explain, if in the first instance I created an endpoint with a single client and then re-write the chip but this time in that same endpoint number I created a client and a server, then the chip was rebooted, I think the reboot occurred because it was trying to create an endpoint that already existed, for example if you try to create an endpoint that already exists by calling the following line 2 times it produces the same reboot
```
esp_zb_ep_list_add_ep(esp_zb_ep_list, esp_zb_cluster_list, ZIGBEE_BOILER_ENDPOINT, 260, ESP_ZB_HA_CUSTOM_ATTR_DEVICE_ID);
esp_zb_ep_list_add_ep(esp_zb_ep_list, esp_zb_cluster_list, ZIGBEE_BOILER_ENDPOINT, 260, ESP_ZB_HA_CUSTOM_ATTR_DEVICE_ID);

```
In summary I think I was not cleaning the endpoint before recording the same endpoint with a different configuration, anyway this problem was fixed after cleaning the memory so i think its not a big issue

**Comment by xieqinan** [comment]
> I am interested in bring the bindings table to confirm if there are duplicates,

The feature has already been added in [esp-zigbee-sdk](https://github.com/espressif/esp-zigbee-sdk/blob/main/components/esp-zigbee-lib/include/zdo/esp_zigbee_zdo_command.h#L399C1-L399C5)



> In summary I think I was not cleaning the endpoint before recording the same endpoint with a different configuration

Good job!
